compiler/qsc_ast/src/ast.rs

```diff
@@ -1968,3 +1994,129 @@ impl ImportOrExportItem {
         }
     }
 }
+
+/// A [`TypeParameter`] is a generic type variable with optional bounds (constraints).
+#[derive(Default, PartialEq, Eq, Clone, Hash)]
+pub struct TypeParameter {
+    /// Class constraints specified for this type parameter -- any type variable passed in
+    /// as an argument to these parameters must satisfy these constraints.
+    pub constraints: ClassConstraints,
+    /// The name of the type parameter.
+    pub ty: Ident,
+    /// The span of the full type parameter, including its name and its constraints.
+    pub span: Span,
+}
+
+impl WithSpan for TypeParameter {
+    fn with_span(self, span: Span) -> Self {
+        Self {
+            span,
+            ty: self.ty.with_span(span),
+            ..self
+        }
+    }
+}
+
+impl TypeParameter {
+    /// Instantiates a new `TypeParameter` with the given type name, constraints, and span.
+    #[must_use]
+    pub fn new(ty: Ident, bounds: ClassConstraints, span: Span) -> Self {
+        Self {
+            ty,
+            constraints: bounds,
+            span,
+        }
+    }
+}
+
+impl std::fmt::Debug for TypeParameter {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        // 'A: Eq + Ord + Clone
+        write!(
+            f,
+            "{}{}{:?}",
+            self.ty,
+            if self.constraints.0.is_empty() {
+                ""
+            } else {
+                ": "
+            },
+            self.constraints
+        )
+    }
+}
+
+/// A list of class constraints, used when constraining a type parameter.
+#[derive(Default, PartialEq, Eq, Clone, Hash)]
+pub struct ClassConstraints(pub Box<[ClassConstraint]>);
+
+/// An individual class constraint, used when constraining a type parameter.
+/// To understand this concept, think of parameters in a function signature -- the potential arguments that can
+/// be passed to them are constrained by what type is specified. Type-level parameters are no different, and
+/// the type variables that are passed to a type parameter must satisfy the constraints specified in the type parameter.
+#[derive(Default, PartialEq, Eq, Clone, Hash, Debug)]
```

Comment:
I've been trying to squash `Default` impls for the AST nodes because I think they also lead to bugs. I haven't been successful in eliminating all of them, but it'd be nice if we can stop adding more. It's too easy to end up with `""` idents in the AST, and then have the resolver spew nonsense errors like "`""` is not a name` etc. I think the compiler code would just be easier to work with if we stuck to invariants such as "no empty string identifiers". 

This one seems to be unused, so I think you can just delete it.

Reply:
Makes sense!